File: packages/ballerina-central/src/render/typedef.ts
/*
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/**
 * One module-level declaration, as Ballerina.
 *
 * `renderTypeDef` switches over `kind` with a `never` fallthrough: adding a
 * Ballerina shape to the IR fails the build here until it has a rendering. It is
 * also the whole of the `type` verb's output — that verb selects declarations
 * and prints them, so a declaration read by name is byte-identical to the same
 * declaration inside the API document.
 *
 * Spacing differs by kind and is load-bearing: an undescribed record still opens
 * with a blank line and an undescribed enum does not. Neither is an accident,
 * and changing either moves every snapshot.
 */

import type { TypeDef } from "../model.js";
import {
  applyPrefixToTypeName,
  buildSpecialAgentNote,
  collectExternalLinks,
  renderDescription,
} from "./signature.js";

function renderRecord(typeDef: Extract<TypeDef, { kind: "record" }>): string {
  const lines = [renderDescription(typeDef.description), `type ${typeDef.name} record {`];
  for (const field of typeDef.fields) {
    const links = collectExternalLinks(field.type);
    const typeName = applyPrefixToTypeName(field.type.name, links);
    const optional = field.optional ? "?" : "";
    const defaultValue = field.default === undefined ? "" : ` = ${field.default}`;
    const description = field.description ? `    # ${field.description}\n` : "";
    lines.push(`${description}    ${typeName} ${field.name}${optional}${defaultValue};${buildSpecialAgentNote(links)}`);
  }
  lines.push("};");
  return lines.join("\n");
}

function renderEnum(typeDef: Extract<TypeDef, { kind: "enum" }>): string {
  const members = typeDef.members.map((member) => member.name).join(",\n    ");
  // Joined with "" rather than "\n": renderDescription already ends in a
  // newline, and an enum with no description must not gain a leading blank line.
  return [renderDescription(typeDef.description), `enum ${typeDef.name} {\n    ${members}\n}`].join("");
}

function renderUnion(typeDef: Extract<TypeDef, { kind: "union" }>): string {
  const description = renderDescription(typeDef.description);
  if (typeDef.members.length === 0) return `${description}type ${typeDef.name};`;
  return `${description}type ${typeDef.name} ${typeDef.members.map((member) => member.name).join("|")};`;
}

function renderConstant(typeDef: Extract<TypeDef, { kind: "constant" }>): string {
  const value = typeDef.varType.name === "string" ? `"${typeDef.value}"` : typeDef.value;
  return `${renderDescription(typeDef.description)}const ${typeDef.varType.name} ${typeDef.name} = ${value};`;
}

/**
 * An error declaration, as the four combinations of the two facts Central
 * publishes about it:
 *
 *   distinct + base   `type SslError distinct ClientError;`
 *   distinct only     `type Error distinct error;`
 *   base only         `type X A|B|C;`
 *   neither           `type X error;`
 *
 * Every one of `ballerina/http`'s 56 errors used to render as the last line,
 * which made the subtype hierarchy — and therefore `e is http:ClientRequestError`
 * — unlearnable from the document. The absent-base default is `error` rather
 * than nothing because an error at the top of its own hierarchy narrows the
 * language's `error`, and that is what its declaration says.
 */
function renderError(typeDef: Extract<TypeDef, { kind: "error" }>): string {
  const links = collectExternalLinks(typeDef.base);
  const base = typeDef.base === undefined ? "error" : applyPrefixToTypeName(typeDef.base.name, links);
  const distinct = typeDef.isDistinct ? "distinct " : "";
  return (
    `${renderDescription(typeDef.description)}type ${typeDef.name} ${distinct}${base};` + buildSpecialAgentNote(links)
  );
}

export function renderTypeDef(typeDef: TypeDef): string {
  switch (typeDef.kind) {
    case "record":
      return renderRecord(typeDef);
    case "enum":
      return renderEnum(typeDef);
    case "union":
      return renderUnion(typeDef);
    case "constant":
      return renderConstant(typeDef);
    case "class":
      return `${renderDescription(typeDef.description)}class ${typeDef.name} {\n}`;
    case "error":
      return renderError(typeDef);
    case "other":
      // Named but not rendered: the declaration exists and the caller may need
      // to reference it, so saying so beats omitting it silently.
      return `// Unknown type: ${typeDef.name}`;
    default: {
      const exhaustive: never = typeDef;
      return exhaustive;
    }
  }
}

File: packages/bal-library-tool/native/src/main/java/io/ballerina/library/model/TypeDef.java
/*
 *  Copyright (c) 2026, WSO2 LLC. (http://www.wso2.com)
 *
 *  WSO2 LLC. licenses this file to you under the Apache License,
 *  Version 2.0 (the "License"); you may not use this file except
 *  in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

package io.ballerina.library.model;

import java.util.List;
import java.util.Optional;

/**
 * A module-level type declaration.
 *
 * <p>The renderer switches over these cases with no {@code default}, so a Ballerina shape nobody
 * renders is a compile error rather than a silently dropped declaration.
 *
 * <p>{@link Alias} carries every one of Central's seventeen alias categories, because a Ballerina type
 * alias is one shape however Central files it, and {@link ObjectDef} carries its four object categories for
 * the same reason. A name whose descriptor the reader could not encode is an {@code Alias} with an empty
 * type — it renders as a comment naming itself. There used to be a separate {@code Other} case for that,
 * produced by one patch injecting one name no oracle declares; when the injection went, so did the case,
 * since two spellings of "no descriptor" is one more than the renderer needs.
 *
 * @since 0.1.0
 */
public sealed interface TypeDef {

    String name();

    String description();

    /**
     * A record. {@code isClosed} is {@code record {| |}} — the form that REJECTS extra fields, and the one
     * fact a caller building a value needs, since an open record accepts anything and a closed one does not.
     */
    record Rec(
            String name,
            String description,
            boolean isClosed,
            boolean isDeprecated,
            List<RecordField> fields) implements TypeDef {

        public Rec(String name, String description, List<RecordField> fields) {
            this(name, description, false, false, fields);
        }

        /**
         * The same record with different members.
         *
         * <p>Exists so a patch cannot lose {@code isClosed} by rebuilding: two of them did, and an inclusive
         * record descriptor holding a rest field is a syntax error, so the loss showed up as a document that
         * stopped compiling at exactly the declaration a patch had rewritten.
         */
        public Rec withFields(List<RecordField> replacement) {
            return new Rec(name, description, isClosed, isDeprecated, replacement);
        }
    }

    /**
     * An enum, whose members carry a description each.
     *
     * <p>A member is a name and a description rather than a bare name, because Central publishes both and 13 of
     * the corpus's 65 members are described — including the two that say which PostgreSQL logical-decoding
     * plugin is the standard one. What Central publishes for NO member is its VALUE: 10 of postgresql's 14 have
     * an explicit one and 6 of those differ from the member name in more than case, so
     * {@code enum SSLMode { … VERIFY_CA }} is a legal declaration that is not quite the real one. The
     * description is the only part of that gap either page can close.
     */
    record Enumeration(String name, String description, List<Member> members) implements TypeDef {

        public record Member(String name, String description) { }

        public List<String> memberNames() {
            return members.stream().map(Member::name).toList();
        }
    }

    /**
     * A type alias: a name bound to a type descriptor.
     *
     * <p>ONE case for all seventeen categories Central models separately — string, integer, decimal, boolean,
     * simple-name-reference, array, union, intersection, anydata, any, tuple, function, typedesc, map,
     * stream, table and xml. They differ in nothing a declaration shows: {@code type Id int|string;} and
     * {@code type TsDef string;} are the same shape with different right-hand sides, and Central already
     * publishes each one's descriptor in full.
     *
     * <p>{@code type.name()} empty means the reader could not encode the descriptor. That still renders the
     * name, as a comment, because the declaration exists and a caller may need to know it does.
     */
    record Alias(String name, String description, TypeRef type) implements TypeDef { }

    record Constant(String name, String description, String value, TypeRef varType) implements TypeDef { }

    /**
     * A module-level {@code public final} variable: a value the package computes once and exports.
     *
     * <p>Distinct from {@link Constant}, which is {@code const} — a compile-time constant whose value is part
     * of its type. These are not: {@code public final readonly & Continue CONTINUE = {};} has a value the
     * package builds, so its initialiser is not something a caller can be given. Central publishes 64 of them
     * across the corpus under {@code variables}, and they were parsed and rendered nowhere, so
     * {@code http:CONTINUE} — which compiles from another module, measured — appeared in no verb.
     *
     * @param initialiser the default Central publishes, which for most of these is {@code {}} and is therefore
     *     not the value; {@link Defaults} decides whether it is writable
     */
    record Variable(String name, String description, TypeRef varType, String initialiser, boolean isReadOnly)
            implements TypeDef { }

    /**
     * A class, an object type, a service type or a listener — the four categories Central publishes with one
     * shape, and one IR case, because a Ballerina object declaration is one shape too.
     *
     * <p>{@link Form} and {@link Role} are orthogonal and both load-bearing. {@code Form} is whether the
     * declaration is instantiable: a {@code class} is, an {@code object} type is a contract that is not, and
     * printing the one as the other told a caller to {@code new} an abstract type. {@code Role} is whether its
     * methods are reached with {@code ->} or with {@code .}, which is the difference between
     * {@code db->query(q)} compiling and not.
     *
     * <p>Neither is a flag Central sets. {@code isClient} does not exist in the payload and {@code isService}
     * exists but is false on all 230 objects in the corpus, so both are DERIVED, and derived from the grammar
     * rather than guessed: a {@code remote} method is only legal in a client or service object, and a service
     * type is a service object by definition — which is what {@code http:Service} and {@code kafka:Service}
     * are declared as in their own sources.
     *
     * <p>{@code methods} is Central's {@code methods} array and nothing else. Its {@code otherMethods},
     * {@code lifeCycleMethods} and {@code initMethod} are all SUBSETS of it — measured across every object in
     * the corpus, with no exception — so they are groupings of one list rather than four lists, and reading
     * them all would print {@code init} twice.
     */
    record ObjectDef(
            String name,
            String description,
            Form form,
            Role role,
            boolean isDistinct,
            boolean isReadOnly,
            boolean isIsolated,
            boolean isDeprecated,
            List<RecordField> fields,
            List<Fn> methods) implements TypeDef {

        /** {@code class X { }} versus {@code type X object { };} — instantiable versus a contract. */
        public enum Form { CLASS, OBJECT_TYPE }

        /** How the object's methods are called: plain, {@code ->} on a client, or attached as a service. */
        public enum Role { PLAIN, CLIENT, SERVICE }

        public ObjectDef(String name, String description) {
            this(name, description, Form.CLASS, Role.PLAIN, false, false, false, false, List.of(), List.of());
        }
    }

    /**
     * An error declaration, with the facts that make error handling learnable from the document instead
     * of guessable.
     *
     * <p>Central publishes one key, {@code detailType}, for two different things, and which one it is
     * decides the syntax:
     *
     * <pre>
     *   category "errors"   supertype       type SslError distinct ClientError;
     *   category "records"  detail record   type FHIRServerError distinct error&lt;FHIRServerErrorDetails&gt;;
     * </pre>
     *
     * <p>{@code base} therefore cannot be named for the wire, and {@code detailRecord} is the flag that
     * says which reading applies. This used to be a single field documented as "never a detail record",
     * generalised from nine fixtures in which every {@code detailType} happens to be an error. It is not
     * a safe generalisation: {@code ballerinax/health.clients.fhir} publishes records, and the document
     * printed {@code distinct FHIRServerErrorDetails} — which is not an error type, does not compile as
     * one, and quietly contradicts the promise that a signature from here is the source for what does.
     */
    record ErrorDef(
            String name, String description, boolean isDistinct, Optional<TypeRef> base, boolean detailRecord)
            implements TypeDef {

        /** An error whose {@code base}, if any, is a supertype rather than a detail record. */
        public ErrorDef(String name, String description, boolean isDistinct, Optional<TypeRef> base) {
            this(name, description, isDistinct, base, false);
        }
    }
}
